### PR TITLE
Fix the TestPullImageFromCentralRegistry to use local registry 

### LINKED ISSUE #10966: [integration-cli] [--net none] Fix TestPullImageFromCentralRegistry
Fix TestPullImageFromCentralRegistry so it will still be a functional test, but will pass when the test container has --net none.

Use the local registry in the container to pull an image. Change the name of the test to reflect it is no longer pulling from the central registry.

Error output when tests are run with --net none:

```
--- FAIL: TestPullImageFromCentralRegistry (0.13s)
	docker_cli_pull_test.go:92: pulling the hello-world image from the registry has failed: Pulling repository hello-world
		time="2015-02-11T23:21:38Z" level="fatal" msg="Get https://index.docker.io/v1/repositories/library/hello-world/images: dial tcp: lookup index.docker.io on 8.8.8.8:53: dial udp 8.8.8.8:53: network is unreachable" 
		, exit status 1

```

### PR DESCRIPTION
**Pull request by brahmaroutu**
Closes #10966
Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
Thanks for the commit @brahmaroutu ! This test actually tests hitting the public Docker Hub as the test case name also says. We should definitely have a test that pulls from local private registry instance however we must keep this test and create a new test requirement called `Networking` in requirements.go and add `testRequires(t, Networking)` here to skip this test when `--net=none`.

cc: @jfrazelle @tianon

**Comment by jfrazelle** [comment]
I dont think we should skip it i think we should also run a v1 registry
locally

**Comment by ahmetalpbalkan** [comment]
@jfrazelle test case actually tests hitting docker hub: TestPullImageFrom**CentralRegistry**. Are we going to change the definition of the test case? If we do that, no tests will remain testing "pull from docker hub" case.

If we run v1/v2 and "docker pull example.com:5000/repo/image", we're no longer testing "docker pull hello-world". These are different things, right?

It makes more sense to keep this as is and skip in `--net=none` case.

@brahmaroutu this PR probably should include TestPullNonExistingImage as well. That does the same thing except expects failure.

**Comment by brahmaroutu** [comment]
@jfrazelle I can add local v1 registry but still the naming on the test is an issue. If a customer sees the test passing when actually he does not have external connectivity it will cause unnecessary alarms. Few options as suggested by @estesp and @ahmetalpbalkan, we skip or we use a conditional. Someone running tests with --net=none should voluntarily set the conditional to have the tests pass?
Please advice, in the mean time I will work on getting a local v1 registry like v2?
@ahmetalpbalkan TestPullNonExistingImage addresses your concern about coverage, I will do that separately.

**Comment by jfrazelle** [comment]
we dont want to rely on the hub

On Thu, Feb 26, 2015 at 1:24 PM, Srini Brahmaroutu <notifications@github.com

**Comment by jfrazelle** [comment]
@tianon correct me if I am wrong the point of --net none is not to skip
tests but to make sure we only use what is in the container, relying on the
hub causes a lot of false failures when it is down

**Comment by ahmetalpbalkan** [comment]
@jfrazelle I totally get that but isn't "pull from docker hub" a valid use case which should be tested?

**Comment by tianon** [comment]
I agree with Jess that this test ultimately ought to go completely, but I
think the reality right now is that it does serve a purpose.  When it's as
"simple" as v1 + v2, testing against two local daemons will be perfect.

Right now, we've got v1, old v2, and new v2, and this test is hitting that
"old v2" branch of code we have no other way to test, right?

**Comment by jfrazelle** [comment]
oh man we have started talking about old v2

ping @dmcgowan

/exits

**Comment by dmcgowan** [comment]
The setupRegistry setups of the most recent version of the v2 registry.  I do agree this test eventually needs to go, likely when the registry code gets migrated to distribution. Trying to setup a local v1 registry does not achieve what is being tested here and IMO not worth the effort. My vote would be to remove or skip on no network.  We already floated the idea of making this test a non-failure when the error is network connectivity related, just currently not enough information is returned in the error to make that distinction.

Not to make things worse but this isn't the only test that relies on pulling an image from the hub.

**Comment by dmp42** [comment]
So:

 * we already have tests that run against a local registry, so, making this test point to local doesn't add value
 * there are other tests that require network access and that will fail, so, this patch is probably incomplete
 * we should probably use --net none to not run these if there is no network

@brahmaroutu what do you think?

**Comment by ahmetalpbalkan** [comment]
@dmp42 agreed to point 3: we can have testRequires(t, Network) for such tests hitting central registry like this.

**Comment by brahmaroutu** [comment]
@dmp42 1. We need tests that exercise local V1 registry, plan is to change tests appropriately.
2 & 3. I agree that we need a way to skip some tests under certain criteria (ex: n/w availability) It would be nice to print like  "[SKIPPED]: pull - pull images from central repo" or just move network tests under a separate category, test-integration-cli-network. I prefer the first.

**Comment by dmp42** [comment]
@ahmetalpbalkan 
@brahmaroutu  

Bundling a v1 (python) registry into the tests is probably not going to fly (if only, because V1 is on the way out, also, because the hub needs to be tested, and unfortunately you can't reproduce the interactions with it simply with a local registry).
Totally agree on the idea of having a way to disable certain tests when there is no wide network (+ some output) - that would probably short term fix things.

Thanks for your work on this.

**Comment by brahmaroutu** [comment]
@ahmetalpbalkan @dmp42 Please see this updated PR, I detect network and use variation of testRequires() to print the current test as SKIPPED if network is unreachable. I have also added a local test to exercise using V1 registry.

File: integration-cli/requirements.go [hunk 1]
```diff
@@ -88,3 +96,14 @@ func testRequires(t *testing.T, requirements ...TestRequirement) {
 		}
 	}
 }
+
+// testRequires checks if the environment satisfies the requirements
+// for the test to run or skips the tests.
+func testRequiresWithLog(t *testing.T, msg string, requirements ...TestRequirement) {
```
Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: this should not be necessary as `go -test.v` shows it's being skipped.

File: integration-cli/requirements.go [hunk 2]
```diff
@@ -32,6 +33,13 @@ var (
 		func() bool { return supportsExec },
 		"Test requires 'docker exec' capabilities on the tested daemon.",
 	}
+	Network = TestRequirement{
+		func() bool {
+			_, err := http.Get("http://google.com")
```
Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: I've mixed feeling about this. Can we `ifconfig` in --net=none container and figure this out from the interfaces? like ifconfig output contains "eth0" or something?

Also at least we should cache this result.

File: integration-cli/requirements.go (hunk 1 above)
Comment (brahmaroutu) [inline integration-cli/requirements.go]: @ahmetalpbalkan The reason I do this is that the output looks cleaner (make test) and we can see Skipped tests like below...
[PASSED]: pull - pull fooblahblah1234 (non-existing image)
[SKIPPED]: pull - pull official names
[PASSED]: pull - pull images from local repo

File: integration-cli/requirements.go (hunk 2 above)
Comment (brahmaroutu) [inline integration-cli/requirements.go]: Yep, When you said no env variables, I thought this is one of the ways to detect network availability, I am open to suggestions.  I can also also use ping onto Docker hub?

Comment (tianon) [inline integration-cli/requirements.go]: IMO, this is reasonable, but it's really testing for "internet
availability" not "network access" (which is the real heart of what we want
to gate these "network" tests on, right?)

Comment (brahmaroutu) [inline integration-cli/requirements.go]: @tianon I agree that we should be sparingly use env variables. But in this case, why can't we have a env to run tests in no network mode?

Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: I still don't get why you don't want to make use of `ifconfig`?

```sh
➜  ~  docker run --net=none busybox ifconfig
lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

➜  ~  docker run busybox ifconfig
eth0      Link encap:Ethernet  HWaddr 02:42:AC:11:00:02
          inet addr:172.17.0.2  Bcast:0.0.0.0  Mask:255.255.0.0
          UP BROADCAST  MTU:1500  Metric:1
          RX packets:2 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:180 (180.0 B)  TX bytes:90 (90.0 B)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
```

File: integration-cli/requirements.go (hunk 1 above)
Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: I'm -1 for this sorry, not only you're duplicating the t.Done() message, you're now missing messages for other `t.Skip`s and other calls to `testRequires`. Either fix them all or fix none, and this really doesn't add any value IMO as it's really obvious in `-test.v`.

Also this messes up `-test.v` output by:

```
[PASSED]: pull - pull fooblahblah1234 (non-existing image)
--- SKIP: TestPullOfficialNames (0.00s)
	requirements.go:87: Actual reason why we fix this test so next line is redundant
[SKIPPED]: pull - pull official names
[PASSED]: pull - pull images from local repo
```

Comment (brahmaroutu) [inline integration-cli/requirements.go]: Few things are not clear yet to me. I am ready to change this code or refactor and merge this into the existing method call but would like to see few more opinions.
What happens when I simply do "make test" ? Do I see clean output saying SKIP: ...?
Secondly, I advocate we have [PASSED], [SKIPPED] or [NOTIMPLEMENTED or something if go permits] for all tests and that is easy to track using tools?

File: integration-cli/requirements.go (hunk 2 above)
Comment (brahmaroutu) [inline integration-cli/requirements.go]: sure, are we guaranteed to always count on eth0?

Comment (tianon) [inline integration-cli/requirements.go]: Not at all -- if we use "--net=host", for example, vanity network interface
names are a very real possibility (like "lan0" here on my local box).

File: integration-cli/requirements.go (hunk 1 above)
Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: @brahmaroutu you don't see SKIP: in `make test` because SKIP is something that should be shown in a more verbose level. If you have a problem with that, probably you need to talk to golang developers as they made the design decision.

If you run the tests like `TESTFLAGS='-test.v' make test` yes you will see the clean output saying `SKIP:` and that's what the `t.Skip(r.SkipMessage)` is supposed to do with the `r.Message`. I don't think we should be reinventing the wheel here and just stick with the framework defaults (in this case pkg/testing).

File: integration-cli/requirements.go (hunk 2 above)
Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: @tianon then we can count on the number of interfaces being `>0` if loopback is always going to be there.

**Comment by brahmaroutu** [comment]
Any ideas on how network should be detected or we live with this in this iteration.
Also I prefer running tests without test.v option, I still think that has cleaner output. 
There are few PRs hung up on this decision.  Please comment.

**Comment by ahmetalpbalkan** [comment]
@brahmaroutu 

1. as I said we can probably look at ifconfig output for interfaces that are not `lo` (https://github.com/docker/docker/pull/11048#discussion_r260887040 

2. I'm -1 for `testRequiresWithLog` (https://github.com/docker/docker/pull/11048#discussion_r26087546) we simply don't need it. In the latest windows CI tests, we started to get green but we skip at least 20-30 test cases and we removed `-test.v` because they're not really necessary. anyone who can running tests on their machine can see the skips easily with `-test.v`.

**Comment by dmcgowan** [comment]
Can you update `TestPullImageFromCentralRegistry` in #11264 and close this PR?  Otherwise this can be merged after rebasing on the other PR with the v1 registry stuff removed.

**Comment by dmcgowan** [comment]
Thank you for rebase, can you also please fix to only include the update in `TestPullImageFromCentralRegistry`.  Pulling in the local v1 registry is no longer needed.

**Comment by icecrime** [comment]
Carried in #11729.
